### DISCUSSION TITLE
Atop

### DISCUSSION THREAD

**Discussion by MirceaKitsune** [redeclipse/base#1366] [Maps and Content Showcase]
My first ever map for RedEclipse. Not fully satisfied with a few things and mistakes that may have been made, but for a one week project I'm definitely happy with the end result, especially when you consider how 10 days ago I still had no idea how the editor works and was learning how you drag those cubes around.

This is a medium sized urban map where the action takes place atop three tall buildings with helipads. It was inspired by UT's Morpheus though is structurally quite different, and probably the first low gravity map made for RE. I tried going for a lot of detail and seeing what the engine can do, might have pushed a few limits and some breakage is possible.

My goal is to get the map included but it's likely not there just yet. Please test it and let me know what you think or what issues you find. Only explicit problem I'm aware of is that jump pads will sometimes launch bots in a different direction, I can't replicate this as a player so no idea. I'm a bit worn out after working so much on this, and since I'm technically still a beginner I wouldn't mind if someone else wants to add anything and help finish things up, though I'd be happy if the basics and stuff like the environment was left as is since I tried hard to get the shadows and such to line up right and look awesome. Here are a few screenshots and the zip for the latest version.

[atop_r1997.zip](https://github.com/redeclipse/base/files/14152160/atop_r1997.zip)

![20220918_023115](https://user-images.githubusercontent.com/825418/190879922-f7a89f60-90e3-4d6c-8e8e-04df5855e9d3.png)
![20220918_023821](https://user-images.githubusercontent.com/825418/190879923-ef5e7f11-bafb-488a-b51a-997856aa4739.png)
![20220918_023859](https://user-images.githubusercontent.com/825418/190879925-e073c2b5-6625-4af1-a18d-be825e9879dc.png)
![20220918_024340](https://user-images.githubusercontent.com/825418/190879926-d65d24c6-bab8-4e74-87b3-332fd85636ec.png)
![20220918_023211](https://user-images.githubusercontent.com/825418/190879910-3af23dcc-9001-495d-b0ff-8c13ea2ba9ff.png)
![20220918_023258](https://user-images.githubusercontent.com/825418/190879913-11d3917e-87d6-4273-8a78-b78ebe1b57e7.png)
![20220918_023329](https://user-images.githubusercontent.com/825418/190879914-7f0429f4-c72c-45f8-87be-032669a7dc34.png)
![20220918_023142](https://user-images.githubusercontent.com/825418/190879908-66768b2b-cb63-4e2a-9c5d-43b3b0f1d546.png)

**Comment by Jigoku** [comment]
Really cool! First impressions are very good, interesting environment that looks like it should be a lot of fun to play on. Many little details such as the sawblade cutting into the ship/craft is great, really makes it feel like a workshop etc in there. The blinking red light on one of the helipads is a nice touch as well, small things like that bring a map to life.

Some things that i'd suggest / wonder about; 
* The ship models in some places look very small (although they could be un-manned drones right?) Not a complaint, but just saying :smile: 
![2022-09-18-121619_1920x1080_scrot](https://user-images.githubusercontent.com/1535179/190899404-b8118957-c8ac-41f3-bd7f-82036191a779.png)

* This billboard/sign might be better without the no-parkour material, as it looks like a perfect surface you'd expect to be able to wall run on and get across from one building to another
![2022-09-18-121136_1920x1080_scrot](https://user-images.githubusercontent.com/1535179/190899439-4c5cb969-7580-4d2d-a083-5691665a1b31.png)


* The texture choice here is a bit repetitive, it could have a little more variation (even if some segments are just coloured a little bit darker)
![2022-09-18-121646_1920x1080_scrot](https://user-images.githubusercontent.com/1535179/190899541-34df35ea-5f3c-425a-94c8-5cfbe40549e9.png)
![2022-09-18-120850_1920x1080_scrot](https://user-images.githubusercontent.com/1535179/190899548-3b671d2d-c8fe-4f5a-b0c9-c1a7916e6bf5.png)
as for the above, a simple idea that could add a bit more subtle variation, could be adding some small trim texture to the corners, and alternating some sections with a different colour, or even using a completely different texture on some parts of the structure (so that it isn't one big repeating texture), a quick example that may give some ideas to experiment with:
![2022-09-18-123658_1920x1080_scrot](https://user-images.githubusercontent.com/1535179/190900117-2988bbeb-cd72-4a82-a7a3-8770c1f984ca.png)

Overall it looks very good on both map variants.
The gravity reduction does make it feel really fun, and the environment/setting fits well with the style of the level. Great job so far, especially for a first map. :partying_face:

**Reply by MirceaKitsune** [reply]
Thanks for the feedback! I wanted to give the front of the central tower a ship workshop feel with the back being a bit more like a lab, the sawblade was meant to be a welding machine but I couldn't find a robot arm model so I made that instead. I explicitly wanted a few larger lights at different positions shining at an angle which contrast with the darkness, it makes good use of bump maps and those pretty pixel lighting effects the engine lets us have. Was thinking of adding more detail precisely to those towers but couldn't figure out exactly what to change, I'll keep it on my TODO list for tweaks I'll do later.

The ship models use their original scale, which is a little small by default just as the benches are overly large. I don't think I could resize them though as they would take up too much space and not fit in areas that are already cramped. I was actually thinking of them as one-person ships, though even for one character they'd have to be pretty cramped while driving it 😄

Being able to parkour on the giant sign might be fun. I wanted to exclude that on purpose though: This would make it even easier to quickly get from the tall tower to the bottom one but not the other way around, which could unbalance team gameplay. You can already double-jump to do this which I kind of wanted to avoid, it may be putting the red team at a slight disadvantage. I balanced that by ensuring their sniper tower is taller and more secluded which makes it somewhat easier to hit at a distance without being as easy to hit from the other buildings.

**Comment by MirceaKitsune** [comment]
Next version redoes some of the textures in outdoor areas and will fix the lack of detail on towers. In addition the unnecessary clip volumes surrounding the map were removed, given the player can't jump past the death volume at the bottom and even if they could the engine automatically kills players that reach below height 0 so there's no risk of falling endlessly: You can now parkour on the giant floating sign which you can in fact jump to... there isn't practically much of a point but hey you can!

Will wait for more feedback before posting a new revision in case I should fix any other issues anyone finds. If everything else is good the next version may be the final one. Here's some shots of the new texturing, may still be subject to change.

![20220918_204942](https://user-images.githubusercontent.com/825418/190922079-dc773868-e76e-46a1-9a9d-d5142066708e.png)
![20220918_211724](https://user-images.githubusercontent.com/825418/190922375-b2d17382-daab-40a6-90fb-ee61b44c4a88.png)

**Comment by MirceaKitsune** [comment]
Played a few more local matches and couldn't find any obvious remaining issues, so here's r1253 which is hopefully the final one. In addition to the things mentioned above and a few misc fixes, I did a last minute change to the clouds so they're drawn in the world rather than the skybox which I didn't know you could do, the layers now appear in between buildings which looks a lot more awesome.

![20220918_233750](https://user-images.githubusercontent.com/825418/190935106-4a50c9a1-4caf-4b5b-812e-c3d4dc20a9ba.png)
![20220918_233733](https://user-images.githubusercontent.com/825418/190935105-22d9c4c1-2f20-4613-a9de-373c131d5a8f.png)

Please test and let me know if it plays smoothly. Unless it's important I'm probably not making further changes to this one for now. If no issues are found my goal is to get this in as an official map as the other maps I plan on making.

[atop_r1253.zip](https://github.com/redeclipse/base/files/9595604/atop_r1253.zip)

**Comment by qreeves** [comment]
Alright, welcome to Red Eclipse mapping. This is a great "first map". Most of the issues just come from not knowing the standard practices, so I've gone over the map fairly thoroughly and here's the feedback:

![20220923_171925](https://user-images.githubusercontent.com/3297929/191925688-0cd17699-63f6-402f-8e3b-815b399bc4c0.png)
You've got a few paths for cameras, but they're all in areas where they'd be in the middle of the action all the time. You might want to add some more that create sweeping views of large areas - they don't have to be on rails, you can have static cameras as well. The thing to keep in mind is the spectator system will prioritise cameras that see more players at reasonable distances. There are also some corridors with no cameras so players inside them will pretty much be invisible to the automatic spectator view.

![20220923_172904](https://user-images.githubusercontent.com/3297929/191927050-5db3461e-8ddd-4a77-b60a-dcfe942ac77c.png)
I noticed the envmapped glass windows looked really bad, and it stems from this issue. Each envmap entity creates a static cubemap at that position, and the radius is the maximum distance for surfaces that can reference the cubemap. As the singular envmap entity is in the middle of the level, it shows everything from that perspective leading to really incorrect looking reflections from certain angles on the outlier buildings. If you really want to use envmaps you'll want to add more (but not too many!) or reduce the radius on the single envmap entity to only work on walls in the center area (you can set `/envmapbb 1` to make it a cube instead of a sphere). If there's no reachable envmap it will default to the skybox, which may look more correct elsewhere.

![20220923_173100](https://user-images.githubusercontent.com/3297929/191928075-96782d3a-b2e3-4232-93ee-bf7a26d948c1.jpg)
This is an opinion, but this decal should maybe scaled down a little so it fits inside the dark trim of the texture it's placed on, just think this would look better.

![20220923_175241](https://user-images.githubusercontent.com/3297929/191928317-df6e5609-6286-4e3f-b420-d2ad0fb58da2.jpg)
There's two Defend and Control affinities within a really close distance of each other, I don't generally recommend too many of these in the first place (just one per particular point of interest outside a certain distance of each other).

![20220923_191229](https://user-images.githubusercontent.com/3297929/191928824-43acaf11-eab8-4d13-841e-f59a90df7f9b.png)
Considering you can get outside of the map boundaries pretty easily in this map, the death material is a bit small. I thought about this issue a little and I've changed the hard coded zero point applied to suicide by falling outside the map to a new variable: `deathplane`. You'll be able to see this in editmode as a layer on the map now when `showphyslayers` (PL) is on. I recommend you remove the existing death material at the bottom of the map and set `/deathplane 0.63` and `/deathplanealt 0.63`.

![20220923_173340](https://user-images.githubusercontent.com/3297929/191928975-5e14a37b-b0dd-4ed5-9bc3-22b00dcdaf74.png)
There are issues with this mapmodel because it extends outside the map boundary. I'd probably recommend increasing the map size so players are more likely to stay inside the boundary. It's not a hard rule, but when stuff phases in and out of existence because of the camera being outside the boundary it's something you want to limit as much as possible. You can just clear the geometry that gets created and leave the rest of the level alone in this case as all you need is more vertical space.

**Comment by MirceaKitsune** [comment]
Definitely sounds like I'll need to make a new revision, will keep this open to remember. Thanks for adding it in the meantime: In my tests it plays well so at least in the basics I'm not concerned.

Yes the death material should be removed at this point, I added it because the death height was hardcoded at 0 which was too low and the player would get bored falling, with this change I can get rid of it... can you customize the obituary message for it though? I like how some of the decals use the normal map of the texture they're on, makes them feel like stickers glued onto the surface... some I could make smaller so I don't overdo it I guess. Spectator cameras not sure how much I'll change now, but I see what you mean and should especially keep this in mind for the future. And yeah I made a mistake building so close to the top limit of the map, initially I thought it only mattered in that you can't place entities outside before discovering the additional despawn issues... don't wanna resize so instead I might lower the towers slightly so it's not such a problem.

**Reply by Jigoku** [reply]
>  can you customize the obituary message for it though?
`obitfall` is what you're looking for there

**Reply by Calinou** [reply]
I've tested the version in redeclipse `master` and it's quite good so far :slightly_smiling_face:

My two main concerns are:

- Gravity is very low, which makes gameplay feel floaty. Weapon pushback can easily keep you in the air, which is good when used as a movement tool but not so much when you're just trying to fight while airborne. I'd suggest raising gravity slightly (while still keeping it lower than default).
- Performance suffers in some areas, especially on top of the main tower with the blue-ish volumetric light below the grating. Other demanding maps like `conquest` run significantly faster on my machine. I'd suggest making some lights cast static shadows only, or even no shadows at all for lights where it doesn't make a big difference. Lights that are fairly dim are a good candidate for this.
  - On my setup, reducing shadow map size with `/smsize` is the main way to improve performance, which hints at the fact that too many lights have shadows enabled. `/smsize 10` results in substantially better perforance than `/smsize 11`, which is not the case on most other maps.

**Reply by MirceaKitsune** [reply]
Changing gravity now would break the ability to jump between places as intended. I still feel `/gravity 10` is good for such maps though, against the default of 50: It's a bit floaty as desired but not all that much, higher felt like it's not noticeable enough to make a difference and I wanted it to feel special. Sadly there's no way to set gravity per area so it's only low outside, would have went with that but at the moment it needs to be global.

Didn't notice any performance issues... that is a bit surprising given I used simple geometry and there should be enough occlusion culling for indoor where most detail is. On the other hand I do use quite a few map models, and indeed some volumetric lights too. I tend to dislike lights not casting shadows when that ability exists, but the dim ones in the doorways could totally use static so I'll change that... should really do the same for the lights above weapon glow pads.

**Comment by MirceaKitsune** [comment]
[atop_r1348.zip](https://github.com/redeclipse/base/files/9690606/atop_r1348.zip)

r1348, please update with this in development: Fixes the issues mentioned above and other oversights, less for envmap and camera rails which I looked at but didn't feel I could improve further. The giant antenna has been moved and will no longer pop out of existence if seen from the wrong angle, several lights were made static and / or noshadows which should improve performance. In addition I set a proper skybox behind the atmosphere for yet more detail!

![20221001_171716](https://user-images.githubusercontent.com/825418/193416552-41d28152-1d5b-4b64-9864-f6c57348ecae.png)
![20221001_171754](https://user-images.githubusercontent.com/825418/193416554-27347f56-3713-4bd7-8a21-d22b95468ce9.png)
![20221001_172223](https://user-images.githubusercontent.com/825418/193416437-5dad6abc-d38e-4223-a7d6-78e7c28bb942.png)
![20221001_171803](https://user-images.githubusercontent.com/825418/193416578-fe6ec00e-7662-4f0b-bf89-2d32378b3dc3.png)

**Comment by MirceaKitsune** [comment]
r1357: Mainly fixed the brightness as it was annoyingly dark especially during night time. Edited to include further fixes, please download the latest version from this post again.

[atop_r1357.zip](https://github.com/redeclipse/base/files/9720754/atop_r1357.zip)

**Comment by MirceaKitsune** [comment]
Sorry for the delay in getting this up to date. I have now submitted [r1473](https://github.com/redeclipse/base/files/10710385/atop_r1473.zip) which includes the fixes suggested by @dc-de with better lighting and minor design changes and more. Please update the version in Git master so others can try out the new modifications.

**Comment by MirceaKitsune** [comment]
Submitted [r1953](https://github.com/redeclipse/base/files/13401374/atop_r1953.zip) hotfix: Improves navigation in a several areas... notably replaced the large ship on the tower with a smaller one to free up its helipad, some clutter was optimized to further reduce annoyances you could bump into. A few issues that weren't caught last time were otherwise addressed. Minor visual and texturing updates.

**Comment by MirceaKitsune** [comment]
[r1997](https://github.com/redeclipse/base/files/14152160/atop_r1997.zip) is out: Added the Corroder weapon, low gravity in the exterior area, various small fixes.
